merge serpapi-ruby from dev into master (#1)
serpapi/serpapi-ruby: README.md

# Scrape Google and other search engines from our fast, easy, and complete API using SerpApi.com
#
[![serpapi-ruby](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml/badge.svg)](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml)
[![Gem Version](https://badge.fury.io/rb/serpapi.svg)](https://badge.fury.io/rb/serpapi)

This Ruby Gem is meant to scrape and parse results from all major search engine available world wide including Google, Bing, Baidu, Yandex, Yahoo, Ebay, Apple and more using [SerpApi](https://serpapi.com).
SerpApi.com returns JSON or HTML results. This library does automatically convert JSON in Hash using the json standard library from shipped with Ruby.

SerpApi.com provides a [script builder](https://serpapi.com/demo) to get you started quickly.

## Installation

A modern version of Ruby must be already installed.

```bash
$ gem install serpapi
```

[Link to the gem page](https://rubygems.org/gems/serpapi/)

Tested Ruby versions:
 - 2.2
 - 2.5.2

>  - 3.0.0

Reviewer: Ruby versions in `ci.yml`: 3.0.3, 2.5.2, 2.7.5.

> see: Git hub actions.

Author: fixed.

> see: Git hub actions.

Author: fixed by f886b6ea31f52f028e3b9ddddb4bf7b205bb9da0

  start: "Pagination Offset",
  tbm: "nws|isch|shop",
  tbs: "custom to be client criteria"
  async: true|false # allow async call
}

# formated search results as a Hash
#  serpapi.com converts HTML -> JSON 
results = client.search(query)

# raw html as a String
#  serpapi.com acts a proxy to provive high throughputs
doc = client.html(query)
```

(The full documentation)[https://serpapi.com/client-api].
More hands on examples are available belows.

### Location API

```ruby
client = SerpApi::Client.new() 
location_list = client.location(q: "Austin", limit: 3)
puts "number of location: #{location_list.size}"
pp location_list
```

it prints the first 3 location matching Austin (Texas, Texas, Rochester)
```ruby
[{:id=>"585069bdee19ad271e9bc072",
  :google_id=>200635,
  :google_parent_id=>21176,
  :name=>"Austin, TX",
  :canonical_name=>"Austin,TX,Texas,United States",
  :country_code=>"US",
  :target_type=>"DMA Region",
  :reach=>5560000,
  :gps=>[-97.7430608, 30.267153],
  :keys=>["austin", "tx", "texas", "united", "states"]},
  ...]
```

### Search Archive API

This API allows to retrieve previous client.
To do so run a client to save a search_id.
```ruby

> client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'google')

Author: fixed by 4d279bbe2f0c760c3dc7291772f81a3388f3a837

original_search = client.search(q: "Coffee", location: "Portland")
search_id = original_search[:search_metadata][:id]
```

Now let retrieve the previous client from the archive.

```ruby
client = SerpApi::Client.new(api_key: ENV['API_KEY'])
archive_results = client.search_archive(search_id)
pp archive_results
```
it prints the search results from archive.

### Account API
```ruby
client = SerpApi::Client.new(api_key: ENV['API_KEY'])
pp client.account
```
it prints your account information.

## Basic example per search engine
### Search Google Images
```ruby
require 'serpapi'
client = SerpApi::Client.new(tbm: 'isch', api_key: ENV['API_KEY'], engine: 'google')
image_results_list = client.search(q: 'coffee')[:images_results]

> image_results_list[0..2].each do |image_result|
> puts image_result[:original]

Reviewer: ```suggestion
image_results_list[0..2].each do |image_result|
  puts image_result[:original]
end
```

Author: code update to `pp results[:images_results]`. so it's more generic for all the templated spec files

```
 see: spec/serpapi/example_search_google_images_spec.rb

This code prints the first 3 images links.
To download the image on Linux you could do: `wget #{image_result[:original]}`

### Search youtube
```ruby
require 'serpapi'
client = SerpApi::Client.new(api_key: ENV['API_KEY'])

> query = {
> engine: 'youtube',
> search_query: 'Coffee'
> }

Reviewer: ```suggestion
query = {
  engine: 'youtube',
  search_query: 'Coffee'
}
```

# serpapi-ruby : User Guide

Scrape Google and other search engines from our fast, easy, and complete API using SerpApi.com.

[![serpapi-ruby](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml/badge.svg)](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml)
[![Gem Version](https://badge.fury.io/rb/serpapi.svg)](https://badge.fury.io/rb/serpapi)

SerpApi Gem is meant to scrape and parse results from all major search engines including Google, Bing, Baidu, Yandex, Yahoo, Ebay, Apple and more using [SerpApi](https://serpapi.com).
The service SerpApi.com returns JSON or HTML results. This library does automatically convert JSON in Hash using the json standard library from shipped with Ruby.

## Installation

> A modern version of Ruby must be already installed.

Author: fixed.

```bash

> $ gem install serpapi

Author: done.

```

[Link to the gem page](https://rubygems.org/gems/serpapi/)

Ruby versions validated by Github action:
 - 3.1.2
 - 2.7.5
 - 2.6.9

> see: [Github actions.](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml)

Author: sure. I'll move it.

```ruby
require 'serpapi'
client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'bing')
results = client.search({
  "q": "coffee"
})
pp results[:organic_results]
# ENV['API_KEY'] captures the secret user API available from http://serpapi.com
```

 * source code: [spec/serpapi/example_search_bing_spec.rb](https://github.com/serpapi/serpapi-ruby/blob/master/spec/serpapi/example_search_bing_spec.rb)
 * doc: [https://serpapi.com/bing-search-api](https://serpapi.com/bing-search-api)

### Search baidu
```ruby
require 'serpapi'
client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'baidu')
results = client.search({
  "q": "coffee"
})
pp results[:organic_results]
# ENV['API_KEY'] captures the secret user API available from http://serpapi.com
```

 * source code: [spec/serpapi/example_search_baidu_spec.rb](https://github.com/serpapi/serpapi-ruby/blob/master/spec/serpapi/example_search_baidu_spec.rb)
 * doc: [https://serpapi.com/baidu-search-api](https://serpapi.com/baidu-search-api)

### Search yahoo
```ruby
require 'serpapi'
client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'yahoo')
results = client.search({
  "p": "coffee"
})
pp results[:organic_results]

> # ENV['API_KEY'] captures the secret user API available from http://serpapi.com

Reviewer 2: ```suggestion
# ENV['API_KEY'] captures the secret user API. You can find yours here: https://serpapi.com/manage-api-key
```
